### ISSUE TITLE
[BUG] complains of no UTF-8 while locale *is* UTF-8

### ISSUE THREAD

**Issue by joske** [aristocratos/btop#52]
**Describe the bug**

Running on RHEL 7.7 at work, using the statically compiled binary 1.0.9, it complains of no UTF-8 locale detected. While locale is:
[jdehaes@berx771 Snape]$ locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=

When using the --utf-force flag, btop works perfectly. We have the same on other RHEL 7.x boxes.
I can not compile/debug btop on this machine (as its compilers are too old - even with gcc 8.3 -std=c++20 is not available). On a RHEL 8 system this just works. I can compile/debug there (after installing gcc 10 toolchain), but useless as the problem doesn't occur there. The locale settings are identical on that system.

**To Reproduce**

Get a RHEL 7.7 (probably any 7.x will do) system (via docker?) and copy the btop binary. It will complain although locale is UTF-8.

**Expected behavior**

btop can run with the --utf-force flag.

[jdehaes@berx771 Snape]$ btop
ERROR: No UTF-8 locale detected!
Use --utf-force argument to force start if you're sure your terminal can handle it.

[If applicable, add screenshots to help explain your problem.]

**Info (please complete the following information):**
 - btop++ version: 1.0.9
 - Binary: static binary from release
 - Architecture: x86_64
 - Platform: Linux
 - (Linux) Kernel: 3.10.0-1062.4.1.el7.x86_64
 - Terminal used: mate-terminal
 - Font used: Noto Mono for Powerline Regular

**Additional context**

contents of `~/.config/btop/btop.log` -> nothing there other than no UTF-8 locale detected.

**Comment by daniejstriata** [comment]
I'm getting this too on Ubuntu 20.04.3 LTS and Centos 7.9. I am not seeing this on my Centos 8 servers.

**Comment by aristocratos** [comment]
@joske 
Run `locale -a` to check if you actually have a UTF-8 locale generated.

As @daniejstriata pointed out above, you can generate one with `locale-gen` or something equivalent for your distro.

**Comment by daniejstriata** [comment]
@aristocratos Why would we get the error if the locale is UTF8?

**Comment by aristocratos** [comment]
@daniejstriata 
You wouldn't, and as you pointed out, when you generated one it went away.

The output when running just `locale` is just the environment variables of the locales you want to use.
The output of `locale -a` is the actually available locales.

**Comment by daniejstriata** [comment]
On the Centos host I still get the error.
![image](https://user-images.githubusercontent.com/13314239/135074575-82968308-1942-4c5b-9a92-9a01a9cbb003.png)
Output from locale:
```→ locale
LANG=en_ZA.UTF-8
LC_CTYPE="en_ZA.UTF-8"
LC_NUMERIC="en_ZA.UTF-8"
LC_TIME="en_ZA.UTF-8"
LC_COLLATE="en_ZA.UTF-8"
LC_MONETARY="en_ZA.UTF-8"
LC_MESSAGES="en_ZA.UTF-8"
LC_PAPER="en_ZA.UTF-8"
LC_NAME="en_ZA.UTF-8"
LC_ADDRESS="en_ZA.UTF-8"
LC_TELEPHONE="en_ZA.UTF-8"
LC_MEASUREMENT="en_ZA.UTF-8"
LC_IDENTIFICATION="en_ZA.UTF-8"
LC_ALL=
```
output from locale -a
```
→ locale -a | grep -i en_ZA.utf8
en_ZA.utf8
```
output from `localectl status`
![image](https://user-images.githubusercontent.com/13314239/135075234-a45b8b53-d52e-4212-9b21-4652f0591235.png)

**Comment by aristocratos** [comment]
@daniejstriata 
Try with latest commit, made some changes if LANG locale fails to be set.

**Comment by daniejstriata** [comment]
@aristocratos I have tried to compile statically on Fedora 34 but I get a failure. Dynamic is fine:

![image](https://user-images.githubusercontent.com/13314239/135079998-9ab36d8b-2d02-4faf-b30b-0973e20731d5.png)

I have the development tools installed. Regrettably I'm not sure what I'm missing.

**Comment by aristocratos** [comment]
@daniejstriata 
I'm not very well versed in fedora. But you can look for libstdc++ in your package manager and see if you find any missing packages, libstc++-dev or similar.

Also not sure if you just re-ran the make command just for the ouput, but make sure to `make clean` when changing any compiler flags if your not already doing it.

**Comment by daniejstriata** [comment]
@aristocratos I'll try some more. I already had libstdc++.x86_64  and libstdc++-devel.x86_64 installed.

**Comment by daniejstriata** [comment]
On Fedora 34 I had to install:
```
libstdc++-static
glibc-static
libxcrypt-static

```

**Comment by daniejstriata** [comment]
@aristocratos When I run the statically compiled binary I created just now from the latest git repo on a Centos 7 host it quits without any output/errors but. It still works normally on Centos 8 or the Fedora host I built the binary on.

```
$ git describe
v1.0.9-47-g8296c1a

```
```
→ /usr/local/bin/btop --version
btop version: 1.0.9
```

**Comment by aristocratos** [comment]
@daniejstriata 
Not sure why that is.
But there is new static binaries in v1.0.10 release you can try.

**Comment by daniejstriata** [comment]
@aristocratos I installed the version you compiled and is doing the same thing as mine. The exit code is 1
```
 ██████╗ ████████╗ ██████╗ ██████╗
 ██╔══██╗╚══██╔══╝██╔═══██╗██╔══██╗   ██╗    ██╗
 ██████╔╝   ██║   ██║   ██║██████╔╝ ██████╗██████╗
 ██╔══██╗   ██║   ██║   ██║██╔═══╝  ╚═██╔═╝╚═██╔═╝
 ██████╔╝   ██║   ╚██████╔╝██║        ╚═╝    ╚═╝
 ╚═════╝    ╚═╝    ╚═════╝ ╚═╝      Installer

Installing binary to: /usr/local/bin/btop
Installing doc to: /usr/local/share/btop
Installing themes to: /usr/local/share/btop/themes
File: /usr/local/bin/btop
Setting owner root:root
Setting SUID bit
Installed btop Version v1.0.10.

```
```
(01:09:15) ± [root@server1] ~/server-fixes (master ✓) 
→ btop
(01:09:23) ± [root@server1] ~/server-fixes (master ✓) 
→ echo $?
1
```
--debug also made no difference
```

(01:11:53) ± [root@server1] ~/server-fixes (master ✓) 
→ btop --debug
(01:12:00) ± [root@server1] ~/server-fixes (master ✓) 
→ echo $?
1

```

**Comment by aristocratos** [comment]
@daniejstriata 
If you run with `gdb` again and before you quit out enter `bt` what output do you get?

**Comment by daniejstriata** [comment]
There is no stack.

```
→  gdb -ex run /usr/local/bin/btop
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-120.el7
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /usr/local/bin/btop...done.
Starting program: /usr/local/bin/btop 
[Inferior 1 (process 8447) exited with code 01]
(gdb) bt
No stack.
(gdb) 
```

**Comment by aristocratos** [comment]
@daniejstriata 
Forgot that I moved the signal handler init, so needed to change the function for exiting on errors.
Also made some changes to UTF8 detection in latest commit.
See if that works for you.

**Comment by joske** [comment]
@aristocratos Sorry, was busy day at work yesterday. Yes locales are installed:

[jdehaes@berx771 AMH-4.2]$ locale -a | grep en_US
en_US
en_US.iso88591
en_US.iso885915
en_US.utf8

But I guess you already know by know from the others. Can not build on this machine as the most recent GCC I have is 8.3. @daniejstriata  how did you get working compilers on RHEL 7.x?? They are not in EPEL or RHEL satellite.

**Comment by aristocratos** [comment]
@joske 
Static binary you can try.
[btop.zip](https://github.com/aristocratos/btop/files/7249580/btop.zip)

**Comment by joske** [comment]
@aristocratos Unfortunately, same:

[jdehaes@berx771 tmp]$ chmod 755 btop
2021-09-29 09:08:53
[jdehaes@berx771 tmp]$ ./btop
ERROR: No UTF-8 locale detected!
Use --utf-force argument to force start if you're sure your terminal can handle it.

**Comment by aristocratos** [comment]
@joske 
Any difference if you run `LANG= LC_ALL= ./btop` or `LANG=en_US.utf8 LC_ALL=en_US.utf8 ./btop` ?

**Comment by joske** [comment]
@aristocratos No.

[jdehaes@berx771 tmp]$ LANG= LC_ALL= ./btop
ERROR: No UTF-8 locale detected!
Use --utf-force argument to force start if you're sure your terminal can handle it.
2021-09-29 09:14:40
[jdehaes@berx771 tmp]$ LANG=en_US.utf8 LC_ALL=en_US.utf8 ./btop
ERROR: No UTF-8 locale detected!
Use --utf-force argument to force start if you're sure your terminal can handle it.

**Comment by aristocratos** [comment]
@joske 
Try also `LC_ALL=en_US.UTF-8 ./btop`

**Comment by joske** [comment]
@aristocratos Same:
[jdehaes@berx771 tmp]$ LC_ALL=en_US.UTF-8 ./btop
ERROR: No UTF-8 locale detected!
Use --utf-force argument to force start if you're sure your terminal can handle it.

**Comment by aristocratos** [comment]
@joske 
Ok, relaxed the constraints to no try to set the locale if it's already reporting it's valid. Test with this one
[btop.zip](https://github.com/aristocratos/btop/files/7249813/btop.zip)

**Comment by joske** [comment]
@aristocratos Nope:
[jdehaes@berx771 tmp]$ ./btop
ERROR: No UTF-8 locale detected!
Use --utf-force argument to force start if you're sure your terminal can handle it.

**Comment by aristocratos** [comment]
@joske 
And if you run `LC_ALL=en_US.UTF-8 ./btop` and `LANG=LC_ALL=en_US.UTF-8 LC_ALL=en_US.UTF-8 ./btop` now?

**Comment by joske** [comment]
@aristocratos Both fail:
[jdehaes@berx771 tmp]$ LC_ALL=en_US.UTF-8 ./btop
ERROR: No UTF-8 locale detected!
Use --utf-force argument to force start if you're sure your terminal can handle it.
2021-09-29 09:56:34
[jdehaes@berx771 tmp]$ LANG=LC_ALL=en_US.UTF-8 LC_ALL=en_US.UTF-8 ./btop
ERROR: No UTF-8 locale detected!
Use --utf-force argument to force start if you're sure your terminal can handle it.

**Comment by aristocratos** [comment]
@joske 
Oops, should have been `LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 ./btop`

**Comment by aristocratos** [comment]
@joske 
Can also try the lower case variation again `LANG=en_US.utf8 LC_ALL=en_US.utf8 ./btop`

**Comment by joske** [comment]
@aristocratos same.

**Comment by aristocratos** [comment]
@joske 
Ok, try this one, will show your current locale on fail.
[btop.zip](https://github.com/aristocratos/btop/files/7250045/btop.zip)

**Comment by joske** [comment]
@aristocratos 
[jdehaes@berx771 tmp]$ ./btop
ERROR: No UTF-8 locale detected!
Use --utf-force argument to force start if you're sure your terminal can handle it.
Current locale: C

Strange as locale, when run in terminal returns something else entirely:
[jdehaes@berx771 tmp]$ locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=

**Comment by joske** [comment]
Environment does not have all those LC_ vars, but it does have LANG:
[jdehaes@berx771 tmp]$ env | grep LC
2021-09-29 10:26:05
[jdehaes@berx771 tmp]$ env | grep LANG
LANG=en_US.UTF-8

**Comment by aristocratos** [comment]
@joske 
Another try, if it starts this time check your btop.log for warnings after.
[btop.zip](https://github.com/aristocratos/btop/files/7250324/btop.zip)

**Comment by joske** [comment]
@aristocratos This is even worse ;-) : 
[jdehaes@berx771 tmp]$ ./btop
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
Aborted

snip of btop.log (previous attempt and current):
2021/09/29 (10:23:40) | ===> btop++ v.1.0.10
2021/09/29 (10:23:40) | ERROR: No UTF-8 locale detected!
Use --utf-force argument to force start if you're sure your terminal can handle it.
Current locale: C

2021/09/29 (10:45:38) | ===> btop++ v.1.0.10
2021/09/29 (10:45:38) | WARNING: Failed to set locale en_US.UTF-8 continuing anyway.

**Comment by aristocratos** [comment]
@joske 
That's an easy fix, but now we know why it kept failing on your system, it finds a valid locale but aren't allowed to set it. That you're gonna have to look into yourself. Permission errors I can't do anything about from my end.

This one should have the string error fixed:
[btop.zip](https://github.com/aristocratos/btop/files/7250404/btop.zip)

**Comment by joske** [comment]
@aristocratos This one works (although it complained my terminal was one character too narrow (79x24 by accident). When I resized to 80x24 it worked without any arguments :-).

**Comment by aristocratos** [comment]
@joske 
You can try the running the old release binary as superuser or just `sudo btop`, if that works there's definitvly some weird permissions set for your user. Not sure why setting the terminal locale would be restricted otherwise.

**Comment by joske** [comment]
@aristocratos Same with old with sudo:
[jdehaes@berx771 AMH-4.1]$ sudo /usr/local/bin/btop
ERROR: No UTF-8 locale detected!
Use --utf-force argument to force start if you're sure your terminal can handle it.

RHEL has strange defaults, and our company messes with lots of things in the images used for our VMs (although I disable a lot of those already). But the fact that other users report the same, I'm inclined to think it's RHEL and not our company that messes this up. Running btop under strace did not yield any clues as to what is failing exactly. Any ideas on how to troubleshoot (if you still care as it's working now).

**Comment by aristocratos** [comment]
@joske 
I'd rather not get in to the mess of how different distros handles locales. As long as it's working.
This fix does take away some safety since it ignores failures of setting locale, but at least the warnings should give some idea to whats going wrong in case of weird output.

Will include the fix in release v1.0.11 tonight along with new static binaries.

**Comment by daniejstriata** [comment]
> @aristocratos Sorry, was busy day at work yesterday. Yes locales are installed:
> 
> [jdehaes@berx771 AMH-4.2]$ locale -a | grep en_US en_US en_US.iso88591 en_US.iso885915 en_US.utf8
> 
> But I guess you already know by know from the others. Can not build on this machine as the most recent GCC I have is 8.3. @daniejstriata how did you get working compilers on RHEL 7.x?? They are not in EPEL or RHEL satellite.
@joske I built on my Fedora install.

**Comment by daniejstriata** [comment]
@aristocratos The last zip uploaded worked on my Centos 7 servers.

2021/09/29 (11:52:42) | ===> btop++ v.1.0.10
2021/09/29 (11:52:42) | DEBUG: Starting in DEBUG mode!
2021/09/29 (11:52:42) | INFO: Logger set to DEBUG
2021/09/29 (11:52:42) | WARNING: Failed to set locale en_ZA.UTF-8 continuing anyway.
2021/09/29 (11:52:42) | DEBUG: Setting LC_ALL=en_ZA.UTF-8
2021/09/29 (11:52:42) | INFO: Running on /dev/pts/4
2021/09/29 (11:52:44) | INFO: Quitting! Runtime: 00:00:02

**Comment by aristocratos** [comment]
@joske @daniejstriata 
Fixed in release [v1.0.11](https://github.com/aristocratos/btop/releases/tag/v1.0.11)
